scarlet/observation.py

```diff
@@ -117,33 +214,59 @@ def match(self, model_frame, diff_kernels=None):
                 diff_kernels = fft.Fourier(diff_kernels)
             self._diff_kernels = diff_kernels
 
+        # initialize the filter window
+        self.convolution = convolution
         return self
 
+    @property
+    def convolution(self):
+        """Whether to do a real space or k-space convolution
+        """
+        return self._convolution
+
+    @convolution.setter
+    def convolution(self, value):
+        assert value in ["real", "fft"], "`convolution` must be either 'real' or 'fft'"
+        self._convolution = value
+        if value == "real":
+            coords = get_filter_coords(self._diff_kernels[0])
+            self._convolution_slices = get_filter_bounds(coords.reshape(-1, 2))
+
     def _convolve(self, model):
         """Convolve the model in a single band
         """
-        return fft.convolve(fft.Fourier(model), self._diff_kernels, axes=(1, 2)).image
+        if self.convolution == "real":
+            result = _convolve(model, self._diff_kernels.image, self._convolution_slices)
+        elif self.convolution == "fft":
+            result = fft.convolve(fft.Fourier(model), self._diff_kernels, axes=(1, 2)).image
+        else:
+            raise ValueError("`convolution` must be either 'real' or 'fft', got {}".format(self.convolution))
+        return result
 
-    def render(self, model):
+    def render(self, model, in_frame=True):
```

Comment:
how can the model not be contained in the observation frame? When would that be useful?

Reply:
At the very least, this will be a problem for `LowResObservation`

Reply:
This is actually used in several places, most notably in `display.show_sources`. Passing `in_frame=False` allows us to convolve an image (any image) in the model frame to the observed seeing, for example when we just call `src.get_model()` we no longer get the full model but instead receive the stamp that is contained in `src.bbox`. So we don't want `render` to slice that model, we just want to convolve it with the difference kernel and return the result.

Reply:
This signature is inherited by other observations, for now only `LowRes`, where this strategy may not work. The issue is modularity. Your code assumes something about the kind of `model` that is fed to this method (in particular, that it could come from a boxed source). But the `match` method only matches the full model frame to the observed frame, source boxes are not known or respected.

You've created a secondary mechanism for source-in-box convolution. I actually don't even understand how that works with FFT diff-kernels which have a determined image shape, namely that of the observed image. 
e.g. https://github.com/pmelchior/scarlet/pull/165/files#diff-160912423f14aa02a64f6b86c73d3feaR400
`model` has box shape, but the convolution expects `model_frame shape`. I'm puzzled...

For consistency, that mode would need to be supported for all kinds of observations, only for a minor performance benefit in plotting. I suggest that `display.show_sources` generated full-frame models for the `show_render` option, since `render` is only guaranteed to work with model_frame`, and to remove the `in_frame` option.

Reply:
So I looked into it and there a few things going on here. To make it clear, I've bulleted actions to be taken in my comments below.

The first issue is an oversight on my part when I created the `fft` module last spring. I don't think that I appreciated how expensive padding is computationally in numpy, since it seems like a "trivial" operation (more on this below). So it looks like when I re-wrote `Observation.match` to use `fft.convolve` to create the difference kernel it no longer created them in the `model_frame`. My thinking was (probably) that padding takes a trivial amount of time and in general we won't have two images with a matching shape in k-space, so it seemed ok to create the diff_kernel in PSF space and pad both the diff kernel and the model in real time (since we already had to pad the model anyway). But numpy must be doing something weird under the hood, for example

```
import numpy as np

def pad(arr, newshape):
    newshape = np.asarray(newshape)
    currshape = np.array(arr.shape)
    dS = newshape - currshape
    startind = (dS + 1) // 2
    endind = dS - startind
    pad_width = list(zip(startind, endind))
    return np.pad(arr, pad_width)

def pad2(arr, newshape):
    newshape = np.asarray(newshape)
    currshape = np.array(arr.shape)
    dS = newshape - currshape
    startind = (dS + 1) // 2
    endind = dS - startind
    pad_width = list(zip(startind, endind))
    
    result = np.zeros(newshape, dtype=arr.dtype)
    slices = tuple([slice(start, -end) for start, end in pad_width])
    result[slices] = arr
    return result

arr = np.arange(100).reshape(10, 10)
new_shape = (10000, 10000)

%timeit t1 = pad(arr, new_shape)
%timeit t2 = pad2(arr, new_shape)
```

prints

```
343 ms ± 4.21 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
65 µs ± 1.59 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

which is _much_ more expensive than I would have predicted using the strictly numpy code. I probably expected something more like the `pad2` code I wrote above. Most likely _this_ is why our DFT convolutions are so slow, not the back propagation through the FFTs.

* So I suggest that I re-write https://github.com/pmelchior/scarlet/blob/master/scarlet/fft.py#L40 using the above scheme (which will require writing a direct gradient for autograd).

* It seems outside the scope of this ticket to do more, but we should probably open another issue to create the `diff_kernel` in the model frame again because it will save time and can be done in the existing code base. If you really want I can include it in this ticket as well, but be aware that it might change the behavior of your existing notebooks (at least it will for mine since I've been making use of this).

So now you know why it works to convolve with whatever image we like in render, and I do like that flexibility. I don't like putting this in `display` for two reasons:

1. Right now we are always trimming the model in `render` and it isn't clear to the user this is what's happening. It took me a while to figure out that was going on and I know the codebase.

2. I agree with your point about this not working with all observations, in particular `LowResObservation`, but to me this is all the more reason to put this in display. Only the `Observation` knows whether or not it is valid to project the rendered image into to a given frame and putting this function in `display` puts it on the user to know this.

So I suggest the following:

* I rename the boolean `in_frame` argument to `frame`, where a `Frame` to project the observations into can be passed to the method. If `frame` is `None` or is the same as the `model_frame` then it will slice the model appropriately as it does by default now. However if a different `Frame` is passed to observation, then just like `model_for_frame`, `Observation` will have to use a different convolution (since the proposed changes above will by default pre-pad the difference kernel for the model frame) to project the rendered version of the input image back into the input frame. However, if a `frame` is passed to `LowResObservation` then it will throw a `NotImplementedError` until a later date when we may want to implement a more complicated projection algorithm.

Reply:
Wow: that must be record length for a code review comment!

We should move the conversation about the `in_frame` argument offline because that's a bigger can of worms. I do like the suggestion of implementing the FFT padding directly with the gradients, which is what this PR is supposedly about.

Reply:
And I see: you've already done it. Any noticeable performance gains?

Reply:
Unfortunately not really, maybe a percent or less on a smaller blend. I imagine that for a bigger blend this might become more substantial but for smaller ones this is mostly in the noise.